File: LeetCode/46.全排列.py
# -*- coding: utf-8 -*-
'''
给定一个不含重复数字的数组 nums ，返回其 所有可能的全排列 。你可以 按任意顺序 返回答案。
'''

class Solution(object):
	def permute(self, nums):
		"""
		:type nums: List[int]
		:rtype: List[List[int]]
		"""

		'''
		nums 是选择列表，从中选择一个数，选择列表就少了一个数
		然后向前继续进行排列
		'''

		ans = []

		def do(path, num):
			if not num:
				ans.append(list(path))
				return
			nlen = len(num)
			while nlen > 0:
				nlen -= 1
				val = num.pop(0)
				path.append(val)
				do(path, num)
				num.append(val)
				path.remove(val)
		do([], nums)
		return ans

if __name__ == "__main__":
	s = Solution()
	ans = s.permute([1, 2, 3, 4])
	print len(ans)

File: LeetCode/46.全排列.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
		vector<vector<int> > ans;
		vector<int> path;
		path.reserve(nums.size());
		permute(ans, path, nums, nums.size());
		return ans;
    }

	void permute(vector<vector<int> >& ans, vector<int> &path, vector<int>& nums, const int& size){
		if (size <= 0){
			ans.push_back(path);
			return;
		}
		int i = 0;
		while (i < size){
			int val = nums[i];
			swap(nums[i], nums[size - 1]);
			path.push_back(val);
			permute(ans, path, nums, size - 1);
			path.pop_back();
			swap(nums[i], nums[size - 1]);
			i++;
		}
	}
};